### PR TITLE
zfs root support

### PR DESCRIPTION
**Pull request by Klowner**
This is working for me, I'm in process of writing up instructions for installation

### CODE REVIEW

**Comment by Klowner** [comment]
One thing I've noticed while using this is that when I have zpools which span multiple disks, it will attempt to import all of them, resulting in "pool is already defined" messages at boot. This doesn't cause any problem, but I don't know if it's worth my time to add additional checking to see if the pool already exists.

😕  Input is welcome.

**Comment by jirutka** [comment]
@ncopa could you please review this patch?

**Review by ncopa — CHANGES_REQUESTED** [review]
I think this needs rebase now. sorry.

File: initramfs-init.in [hunk 1]
```diff
@@ -343,6 +343,11 @@ if [ -n "$KOPT_cryptroot" ]; then
 	fi
 fi
 
+if [ "$KOPT_rootfstype" == "zfs" ]; then
```
Comment (ncopa) [inline initramfs-init.in]: use `=` instead of `==`

**Comment by Klowner** [comment]
Will do!

**Review by ncopa — APPROVED** [review]
this lgtm. I mean last commit looks good.

**Comment by Klowner** [comment]
rebased

File: nlplug-findfs.c [hunk 2]
```diff
@@ -496,6 +497,25 @@ static void start_lvm2(char *devnode)
 		spawn_command(&spawnmgr, lvm2_argv, 0);
 }
 
+static void start_zpool(char *devnode)
+{
+	static blkid_cache cache = NULL;
+
+	if (cache == NULL)
+		blkid_get_cache(&cache, NULL);
```
Comment (ncopa) [inline nlplug-findfs.c]: we already have a [blkid cache](https://github.com/alpinelinux/mkinitfs/blob/master/nlplug-findfs.c#L317), which I think will always be initialized when reaching this point. see https://github.com/alpinelinux/mkinitfs/blob/master/nlplug-findfs.c#L912

In any case, we should reuse the same blkid_cache and not try set up a new one. This means that we need pass the cache or maybe even the conf struct as an arg to `start_zpool`

**Review by ncopa — CHANGES_REQUESTED** [review]

File: nlplug-findfs.c [hunk 3]
```diff
@@ -934,8 +954,12 @@ static int searchdev(struct uevent *ev, const char *searchdev, int scanbootmedia
 			start_mdadm(ev->devnode);
 		} else if (strcmp("LVM2_member", type) == 0) {
 			start_lvm2(ev->devnode);
+		} else if (strcmp("zfs_member", type) == 0) {
+			start_zpool(ev->devnode);
 		} else if (scanbootmedia) {
 			rc = scandev(conf, ev->devnode, type);
+		} else if (bootrepos) {
+			rc = find_bootrepos(ev->devnode, type, bootrepos, apkovls);
```
Comment (ncopa) [inline nlplug-findfs.c]: This hunk needs to be removed. `find_bootrepos` got renamed to `scandev` and `bootrepos` got replaced by `scanbootmedia` in 7673ee7d1ce6786a5471ead15513d62027c9756a

**Comment by ncopa** [comment]
Can you please test if this works? It builds at least.

````diff
diff --git a/nlplug-findfs.c b/nlplug-findfs.c
index 52fac0d..3ff2a1d 100644
--- a/nlplug-findfs.c
+++ b/nlplug-findfs.c
@@ -497,23 +497,14 @@ static void start_lvm2(char *devnode)
                spawn_command(&spawnmgr, lvm2_argv, 0);
 }
 
-static void start_zpool(char *devnode)
+static void start_zpool(char *uuid)
 {
-       static blkid_cache cache = NULL;
-
-       if (cache == NULL)
-               blkid_get_cache(&cache, NULL);
-
-       char *uuid = blkid_get_tag_value(cache, "UUID", devnode);
        char *zpool_argv[] = {
                ZPOOL_PATH, "import", uuid,
                NULL
        };
        if (use_zpool && uuid)
                spawn_command(&spawnmgr, zpool_argv, 0);
-
-       if (uuid)
-               free(uuid);
 }
 
 static int read_pass(char *pass, size_t pass_size)
@@ -933,6 +924,7 @@ static int searchdev(struct uevent *ev, const char *searchdev, int scanbootmedia
                blkid_get_cache(&conf->blkid_cache, NULL);
 
        type = blkid_get_tag_value(conf->blkid_cache, "TYPE", ev->devnode);
+       uuid = blkid_get_tag_value(conf->blkid_cache, "UUID", ev->devnode);
 
        if (searchdev != NULL) {
                if (strncmp("LABEL=", searchdev, 6) == 0) {
@@ -940,7 +932,6 @@ static int searchdev(struct uevent *ev, const char *searchdev, int scanbootmedia
                        if (label && strcmp(label, searchdev+6) == 0)
                                rc = FOUND_DEVICE;
                } else if (strncmp("UUID=", searchdev, 5) == 0) {
-                       uuid = blkid_get_tag_value(conf->blkid_cache, "UUID", ev->devnode);
                        if (uuid && strcmp(uuid, searchdev+5) == 0)
                                rc = FOUND_DEVICE;
                }
@@ -955,11 +946,9 @@ static int searchdev(struct uevent *ev, const char *searchdev, int scanbootmedia
                } else if (strcmp("LVM2_member", type) == 0) {
                        start_lvm2(ev->devnode);
                } else if (strcmp("zfs_member", type) == 0) {
-                       start_zpool(ev->devnode);
+                       start_zpool(uuid);
                } else if (scanbootmedia) {
                        rc = scandev(conf, ev->devnode, type);
-               } else if (bootrepos) {
-                       rc = find_bootrepos(ev->devnode, type, bootrepos, apkovls);
                }
        }
 
````

File: initramfs-init.in [hunk 4]
```diff
@@ -363,6 +363,11 @@ if [ -n "$KOPT_nbd" ]; then
 	setup_nbd || echo "Failed to setup nbd device."
 fi
 
+if [ "$KOPT_rootfstype" = "zfs" ]; then
+	# zpool reports /dev/zfs missing if it can't read /etc/mtab
+	ln -s /proc/mdstat /etc/mtab
```
Comment (ncopa) [inline initramfs-init.in]: shouldn't the symlink point to `/proc/mounts`? In any case, it sounds like an upstream bug.

**Comment by Klowner** [comment]
I'll see if this works, sorry for the need for all the changes.

**Comment by Klowner** [comment]
@ncopa Included your suggestions/changes and tested yesterday, works great for me! :+1:

**Pull request closed without merging**